### ISSUE TITLE
HTML metadata for 'cards'

### ISSUE THREAD

**Issue by mnot** [ietf-tools/xml2rfc#776]
It would be good to add appropriate metadata to the generated HTML so that it shows up in 'cards' when shared, e.g., on social media. For example:
  https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/summary-card-with-large-image

**Comment by martinthomson** [comment]
This might need to be optional in the sense that some people might not want metadata proliferation.  @mnot has linked twitter-specific metadata, but there is a bunch of different options to choose from.  This might be a better reference: https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml/#metadata

**Comment by jrlevine** [comment]
I don't think we want to go down the path of adding random non-standard stuff to our rendered documents. Once Twitter is under new management, who knows what they might or might not support.
If you think this is a good idea, how about writing a postprocessor for the HTML that adds whatever and we can see how it looks and how useful it's likely to be?

**Comment by mnot** [comment]
I can do that. It's not just twitter, of course.

**Comment by mnot** [comment]
Hm, it appears that data tracker is blocking *everything* with robots.txt, so that would need to be changed for this to have any practical effect (at least on that site).

**Comment by jrlevine** [comment]
We have encouraged search engines to prefer the RFCs at rfc-editor.org for a long time. I just got Google Scholar to start indexing it last month, at least the pre-XML ones.

**Comment by mnot** [comment]
Sure - but what about I-Ds?

**Comment by jrlevine** [comment]
I don't think this is the right place to ask that question.

**Comment by mnot** [comment]
OK, quick and dirty test, it appears that Twitter pays attention to [OGP metadata](https://ogp.me/), which is good, because that's also what other social networks use.

I added this to a copy of the bcp56bis I-D:

~~~ html
<meta property="og:title" content="Building Protocols with HTTP" />
<meta property="og:description" content="Applications often use HTTP as a substrate to create HTTP-based APIs. This document specifies best practices for writing specifications that use HTTP to define new application protocols. It is written primarily to guide IETF efforts to define application protocols using HTTP for deployment on the Internet but might be applicable in other situations." />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-bcp56bis" />
<meta property="og:image" content="https://www.ietf.org/media/images/ietf-logo.original.png" />
~~~

And got this with the Twitter validator:

<img width="614" alt="Screen Shot 2022-05-11 at 12 48 14 pm" src="https://user-images.githubusercontent.com/74384/167758866-96047cea-3d95-41ec-b1ff-73b1b21b4f39.png">

A couple of things to work out:

1. The canonical URL will need to be generated from info in the source (status, filename, etc.)
2. The image should probably be cropped to square, as that's most likely to work wherever it pops up
3. The description here is the text from the first paragraph of the abstract with tags stripped. We can probably just include all such text from the abstract; I don't see any length limit in OGP.
4. It's probably worth using a separate logo for I-Ds which aren't adopted by a WG.

**Comment by mnot** [comment]
> I don't think this is the right place to ask that question.

Sure, OK, whatever.

**Comment by rjsparks** [comment]
> Hm, it appears that data tracker is blocking _everything_ with robots.txt, so that would need to be changed for this to have any practical effect (at least on that site).

This is not intended, and will be reworked. (We also need a real sitemap for datatracker).

**Comment by larseggert** [comment]
> It's probably worth using a separate logo for I-Ds which aren't adopted by a WG.

@ghwood, could you prepare some square IETF logos with badges like "adopted work item", "submitted for discussion" or the like?

**Comment by martinthomson** [comment]
How about this one?

![image](https://user-images.githubusercontent.com/67641/170214086-aa17a85b-5d0b-4240-a0d5-277c6be6fa33.png)

**Comment by JayDaley** [comment]
Is there any compelling reason we need to add this metadata (or any other metadata) to the raw XML files for I-Ds and RFCs, rather than to the generated HTML pages?  The HTML seems more appropriate as a target for a card, and of course it is easier for us to change, doesn't run into the the "RFCs can never change rule" (which might bite us if we want to retrofit better metadata later, and of course because metadata is well, meta.

**Comment by jrlevine** [comment]
None at all.  I'm only proposing to add it to the generated HTML, and separately to the HTML for pre-XML RFCs and to the rfc-editor info pages.
There is a separate issue that the `<link>` elements that the preptool adds to the prepped XML are completely different to what the spec says.

**Comment by ghwood** [comment]
> > It's probably worth using a separate logo for I-Ds which aren't adopted by a WG.
> 
> @ghwood, could you prepare some square IETF logos with badges like "adopted work item", "submitted for discussion" or the like?

Definitely. 

It would be helpful to understand some details to ensure badges deliver the desired effect.

For example, would badges be applied to I-Ds submitted for any intended publication stream or just IETF? 

I think badges for I-Ds at the "submitted for discussion" stage should not include the intended stream logo.

If there are other factors to consider, I can work up some example badges.

**Comment by rjsparks** [comment]
This is going to require whatever renders the html to dive deep into datatracker state to _know_ if the draft is in a stream and we'll need to discuss whether "candidate for adoption in a wg" is enough to give it the ietf stream label, or if we have to also look at stream state.

Greg would probably want to also consider how much of the card the logo gets to take, and whether a partial logo similar to what Mark proposed is acceptable.

Is this really what we want? Are these cards the right place to fight the "is this draft really a part of the IETF, and what do you mean by IETF" battle?

**Comment by mnot** [comment]
>  This seems getting rather overengineered. For RFCs it's easy enough to tell what the stream is because it's in the tags, but for I-D's I would avoid trying to be too clever. You could assume draft-ietf-* draft-iab-* draft-irtf-* get logos for ietf, iab, and irtf but I wouldn't go beyond that.

Exactly. For I-Ds without those prefixes, something incredibly generic would be suitable -- maybe something not associated with the IETF at all. Or, just omit it.
